### PR TITLE
Allow bridging of private channels via the provisioner

### PR DESCRIPTION
**Pull request by Half-Shot**
This requires us to determine the identity of the caller in order to check if they are *in* the channel. Luckily, there is an endpoint for that :)

### CODE REVIEW

File: src/Provisioning.ts [hunk 1]
```diff
@@ -150,25 +164,25 @@ export class Provisioner {
...
         }
         const team = await this.main.datastore.getTeam(teamId);
         if (team === null) {
             throw Error("No team token for this team_id");
         }
         const cli = await this.main.clientFactory.getTeamClient(teamId);
         try {
-            const response = (await cli.conversations.list({
+            let types = "public_channel";
+            if (this.main.config.provisioning?.allow_private_channels !== false) {
```
Comment (jaller94) [inline src/Provisioning.ts:178]: double negation

Wouldn't `allow_private_channels === true` or `allow_private_channels` be better and more natural to read?

File: src/Provisioning.ts [hunk 2]
```diff
@@ -87,6 +87,20 @@ export class Provisioner {
         return (currentCount >= this.main.config.provisioning?.limits?.room_count);
     }
 
+    private async determineSlackIdForRequest(matrixUserId, teamId) {
+        const matrixUser = await this.main.datastore.getMatrixUser(matrixUserId);
+        if (matrixUser === null) {
+            throw Error('No users found');
```
Comment (jaller94) [inline src/Provisioning.ts:93]: Nit-pick: User**s**? We're just looking for one user.

File: src/Provisioning.ts [hunk 3]
```diff
@@ -87,6 +87,20 @@ export class Provisioner {
         return (currentCount >= this.main.config.provisioning?.limits?.room_count);
     }
 
+    private async determineSlackIdForRequest(matrixUserId, teamId) {
+        const matrixUser = await this.main.datastore.getMatrixUser(matrixUserId);
+        if (matrixUser === null) {
```
Comment (jaller94) [inline src/Provisioning.ts:92]: I prefer `(!matrixUser)`.
btw. getMatrixUser's signatures don't match but TypeScript doesn't complain. Some implementations return undefined, others null.

Screenshot_2020-05-07_15-12-18

File: src/Provisioning.ts (hunk 1 above)
Comment (Half-Shot) [reply src/Provisioning.ts:178]: Ah, careful here. `allow_private_channels` is considered truthy unless explicitly false. All other values are true.

**Comment by Half-Shot** [comment]
@jaller94 I didn't actually modify this PR yet :D

File: src/Provisioning.ts (hunk 2 above)
Comment (Half-Shot) [reply src/Provisioning.ts:93]: We should also be returning null.

File: src/Provisioning.ts (hunk 3 above)
Comment (Half-Shot) [reply src/Provisioning.ts:92]: Ah, because MatrixUser is any, so essentially types are disabled here.

**Comment by Half-Shot** [comment]
>fakeDatastore still has the more general signature getMatrixUser(userId: string): Promise<any>.

Since it's not implemented yet, let's leave it for when that day comes.

**Pull request merged**